### PR TITLE
Use pod labels for better manageability

### PR DESCRIPTION
**Pull request by sfc-gh-kmakino**
This PR contains the following new capabilities:

- An agent scaler uses `AGENT_TAG` environment variable for the Joshua agent image tag instead of `latest` if available. This prevents mismatching versions between the scaler and the agents.
- On boot, an agent scaler requests all running agents to exit after the current ensemble when `RESTART_AGENTS_ON_BOOT` argument is set. This allows to refresh agents as quickly as possible in case both agent and scaler images are updated.
- Agent pods will have a proper app label as `joshua-agent`
- Each agent writes its current ensemble in the Pod label `ensemble`
```
$ kubectl --namespace joshua get pods -L app,ensemble
NAME                                       READY   STATUS              RESTARTS   AGE   APP            ENSEMBLE
agent-scaler-deployment-86c984887f-vdzpv   1/1     Running             0          45s   agent-scaler
joshua-agent-221118030447-0-57dtx          1/1     Running             0          3s    joshua-agent   20221118-030444-kmakino-f067840ad81ea406
joshua-agent-221118030447-1-mjtdf          1/1     Running             0          2s    joshua-agent   20221118-030444-kmakino-f067840ad81ea406
joshua-agent-221118030447-2-647qb          1/1     Running             0          2s    joshua-agent   20221118-030444-kmakino-f067840ad81ea406
joshua-agent-221118030447-3-clnhq          1/1     Running             0          2s    joshua-agent   20221118-030444-kmakino-f067840ad81ea406
joshua-agent-221118030447-4-h6jgn          0/1     ContainerCreating   0          2s    joshua-agent   20221118-030444-kmakino-f067840ad81ea406
```

### CODE REVIEW

File: joshua/joshua_agent.py
Comment (johscheuer) [inline joshua/joshua_agent.py:39]: os is already imported in the above block.

File: joshua/joshua_agent.py
Comment (johscheuer) [inline joshua/joshua_agent.py:531]:
```suggestion
            hostname = env['HOSTNAME']
```

The name `pod` might be confusing and someone could assume that this is referring to a Pod and not a string.

File: joshua/joshua_agent.py
Comment (johscheuer) [inline joshua/joshua_agent.py:542]:
```suggestion
                if pod_desc.metadata.labels.get('last_test', 'false') == "true":
```

Is there a reason to put this into a try statement instead of just using a default value and checking if the labels are set? Do we actually care about the value or would a check if the key exists be enough?

File: joshua/joshua_agent.py
Comment (johscheuer) [inline joshua/joshua_agent.py:548]: I don't have enough context about joshua, but should we check if the label is already correct instead of doing the patch always?

File: Dockerfile
Comment (johscheuer) [inline Dockerfile:38]: Unrelated to this PR: Personally I would prefer to pin those to a specific version to reduce the risk of surprise.

File: k8s/agent-scaler/agent-scaler.sh
Comment (johscheuer) [inline k8s/agent-scaler/agent-scaler.sh:29]: Doing a proper label selection and using jsonpath output might be the better solution to using grep + awk, e.g. https://kubernetes.io/docs/reference/kubectl/jsonpath.

Was there any issue with the headers? The grep command should filter them out anyway?

**Review by johscheuer — COMMENTED** [review]
I'm not an expert on the joshua setup, but a few comments from my side.

File: joshua/joshua_agent.py
Comment (johscheuer) [inline joshua/joshua_agent.py:540]:
```suggestion
            if pod_desc.metadata.labels['last_test']:
```

With the assumption that `last_test` will always be true if set we could simplify it like this, but i know to little about the setup :)

I haven't worked with the Python Kubernetes client for a while, but could there be a case where a Pod doesn't have labels? If we are sure our Pods always have at lease one label we can ignore that case.

File: k8s/agent-scaler/agent-scaler.sh
Comment (johscheuer) [inline k8s/agent-scaler/agent-scaler.sh:17]:
```suggestion
# if AGENT_TAG is not set through --build-arg,
# use the default agent image and tag
export AGENT_TAG=${AGENT_TAG:-"foundationdb/joshua-agent:latest"}
```

This would be the same correct?

**Review by johscheuer — COMMENTED** [review]
Changes LGTM 👍 It would be great if we would have some tests for the setup (you don't have to add them in this PR, it's more of a general comment).

I'm happy to approve this PR but I would like to get a second pair of eyes on this PR from someone with more context on Joshua.

**Review by ammolitor — APPROVED** [review]

**Review by ammolitor — DISMISSED** [review]

**Comment by ammolitor** [comment]
@sfc-gh-kmakino can we check on why the GitHub Test suite timed out? Do we need to update a test with these changes?

**Comment by sfc-gh-kmakino** [comment]
@ammolitor I fixed the CI issue, please take another look

**Review by ammolitor — APPROVED** [review]

**Pull request merged**
